monkeytype/stubs.py

```diff
@@ -542,13 +545,47 @@ def rewrite_TypedDict(self, typed_dict: type) -> type:
-            self.stubs.extend(stubs)
-        self.stubs.append(ClassStub(name=f'{class_name}(TypedDict)',
-                                    function_stubs=[],
-                                    attribute_stubs=attribute_stubs))
+        required_fields, non_required_fields = field_annotations(typed_dict)
+        has_required_fields = len(required_fields) != 0
+        has_non_required_fields = len(non_required_fields) != 0
+        if not has_required_fields and not has_non_required_fields:
+            raise Exception("Expected empty TypedDicts to be shrunk as Dict[Any, Any]"
+                            " but got an empty TypedDict anyway")
+        elif has_required_fields and not has_non_required_fields:
+            for name, typ in required_fields.items():
+                rewritten_type, stubs = self.rewrite_and_get_stubs(typ, class_name_hint=name)
+                attribute_stubs.append(AttributeStub(name, rewritten_type))
+                self.stubs.extend(stubs)
+            self.stubs.append(ClassStub(name=f'{class_name}(TypedDict)',
+                                        function_stubs=[],
+                                        attribute_stubs=attribute_stubs))
+        elif not has_required_fields and has_non_required_fields:
+            for name, typ in non_required_fields.items():
+                rewritten_type, stubs = self.rewrite_and_get_stubs(typ, class_name_hint=name)
+                attribute_stubs.append(AttributeStub(name, rewritten_type))
+                self.stubs.extend(stubs)
+            self.stubs.append(ClassStub(name=f'{class_name}(TypedDict, total=False)',
+                                        function_stubs=[],
+                                        attribute_stubs=attribute_stubs))
+        else:
+            base_class_name = class_name
+            subclass_name = get_typed_dict_class_name(self._class_name_hint + 'NonTotal')
+            for name, typ in required_fields.items():
+                rewritten_type, stubs = self.rewrite_and_get_stubs(typ, class_name_hint=name)
+                attribute_stubs.append(AttributeStub(name, rewritten_type))
+                self.stubs.extend(stubs)
+            self.stubs.append(ClassStub(name=f'{base_class_name}(TypedDict)',
+                                        function_stubs=[],
+                                        attribute_stubs=attribute_stubs))
+            attribute_stubs = []
+            for name, typ in non_required_fields.items():
+                rewritten_type, stubs = self.rewrite_and_get_stubs(typ, class_name_hint=name)
+                attribute_stubs.append(AttributeStub(name, rewritten_type))
+                self.stubs.extend(stubs)
+            self.stubs.append(ClassStub(name=f'{subclass_name}({make_forward_ref(base_class_name)}, total=False)',
+                                        function_stubs=[],
+                                        attribute_stubs=attribute_stubs))
```

Comment:
It seems this section has more duplicated code than necessary. Why can't we just `if required_fields` process those (and emit the total=True TypedDict), then `if non_required_fields` process those and emit the total=False one (possibly subclassing if we generated one with required fields).

Reply:
Yeah, that was pretty awful duplication. I've extracted the common code. However, I'm keeping the branches so that it is explicit rather than implicit that one of them emits two stubs.